easybuild/easyconfigs/r/RELION/RELION-3.0.7-foss-2019a.eb

```diff
@@ -0,0 +1,68 @@
+easyblock = 'CMakeMake'
+
+name = 'RELION'
+version = '3.0.7'
+
+homepage = 'http://www2.mrc-lmb.cam.ac.uk/relion/index.php/Main_Page'
+description = """RELION (for REgularised LIkelihood OptimisatioN, pronounce rely-on) is a stand-alone computer
+ program that employs an empirical Bayesian approach to refinement of (multiple) 3D reconstructions or 2D class
+ averages in electron cryo-microscopy (cryo-EM)."""
+
+toolchain = {'name': 'foss', 'version': '2019a'}
+toolchainopts = {'openmp': True, 'usempi': True}
+
+source_urls = ['https://github.com/3dem/relion/archive']
+sources = [
+    '%(version)s.tar.gz',
+]
+patches = [
+    '%(name)s-3.0.4_no_which_cmd.patch',
+    '%(name)s-3.0.4_drop_usage_of_mpi_include_path.patch',
+    '%(name)s-3.0.5_add_hpc2n_submit_file_args.patch',  # This patch is a site specific patch. Either adapt or drop.
+    ('relion_sbatch.sh', '.'),  # This submit file template is used by the hpc2n_submit_file patch.
+]
+checksums = [
+    '385ae3ae992cef60aa4ce97075dd4580bf00172154af426adf5ec509e248daac',  # 3.0.7.tar.gz
+    'bfdaf1ce856d20c27a5fcf50252394e81d08259899abc3564f6c7bcc2b4000de',  # RELION-3.0.4_no_which_cmd.patch
+    # RELION-3.0.4_drop_usage_of_mpi_include_path.patch
+    'c135b72ff8d4c3015a3b293fa07a2283396bfb3fd0658d5e938ec33fca4dcebc',
+    '054e382e19f7bfd5f016017189e4132337f7de4dc6bf2996f5933bfade61b2f4',  # RELION-3.0.5_add_hpc2n_submit_file_args.patch
+    '9305f869a07a1beaaa7e3a4670625a045c0cc4762b6b542039123e5a2731023d',  # relion_sbatch.sh
+]
+
+builddependencies = [('CMake', '3.13.3')]
+
+dependencies = [
+    ('X11', '20190311'),
+    ('FLTK', '1.3.5'),
+    ('LibTIFF', '4.0.10'),
+    ('tbb', '2019_U4'),
+    ('ctffind', '4.1.13'),
+    ('MotionCor2', '1.2.6'),
+    ('Gctf', '1.06', '', True),
+    ('gnuplot', '5.2.6'),
+]
+
+configopts = "-DCMAKE_SHARED_LINKER_FLAGS='-lpthread' -DALTCPU=true "
```

Comment:
In our tests we had to add: 
Intel: 
```
configopts += "-DALTCPU=ON -DMKLFFT=ON -DFORCE_OWN_TBB=OFF "
configopts += "-D CMAKE_C_FLAGS='-O3 -ip -g -xCORE-AVX512 -restrict ' -D CMAKE_CXX_FLAGS='-O3 -ip -g -xCORE-AVX512 -restrict'"

Foss:
```
-D CMAKE_C_FLAGS='-O3 -ftree-vectorize -ffast-math -g -march=broadwell -mtune=broadwell -mavx2 -mpclmul -mfsgsbase -mrdrnd -mf16c -mfma -mbmi -mbmi2 -mf16c -mprefetchwt1 ' -D CMAKE_'-O3 -ftree-vectorize -ffast-math -g -march=broadwell -mtune=broadwell -mavx2 -mpclmul -mfsgsbase -mrdrnd -mf16c -mfma -mbmi -mbmi2 -mf16c -mprefetchwt1 '
```

This is according to https://www3.mrc-lmb.cam.ac.uk/relion/index.php/Benchmarks_%26_computer_hardware and the benchmarks seem to confirm it.
to properly support the CPU accelerated mode.

Reply:
EasyBuild by default builds with -march=native. So those flags are not needed.
Whether to use -O3 and -ffast-math is something I haven't investigated yet. But that can easily be turned on using toolchainopts.

Reply:
In our case it was not enough to just add `-DALTCPU=ON` to get proper acceleration. 
We had to add the explicit flags (-xCORE-AVX512/AVX2) to get proper CPU acceleration according to user running some benchmarks (the tests were done for the  for the intel compiler toolchain)
```

- with -xCORE-AVX512: Expectation iteration 26 of 35
- with -xCORE-AVX2: Expectation iteration 23 of 35
- with only DALTCPU=true: Expectation iteration 19 of 35
```
It was compiled on our login node which has the same HW architecture as the compute nodes (Skylakes)

Reply:
Do you have optarch turned off by default or so?

The default with optarch on is -xHOST for intel toolchain and, as mentioned above, -march=native for foss.

Just checked btw, FORCE_OWN_TBB is off by default.
LibTIFF do get picked up correctly.

Reply:
@timeu not sure what your
```
- with -xCORE-AVX512: Expectation iteration 26 of 35
- with -xCORE-AVX2: Expectation iteration 23 of 35
- with only DALTCPU=true: Expectation iteration 19 of 35
```
refers to.

Anyway, I tried the explicit C/CXX flags you mentioned above (foss), and ALTCPU --cpu code is still ~2x slower than without --cpu. It should be the other way around.

Reply:
@akesandgren : AFAIK, we have a pretty standard setup and haven't disabled optarch. I will check the logs. 

Regarding the quoted iterations:
This is bascially a benchmark that a CryoEM user provided us for different configurations. The higher the number the faster the version. The numbers are from the intel version.

Our CryoEM users asked for two relion versions: 
fosscuda (GPU acceeleration)
intel (CPU acceleration).

Reply:
Yeah, but there is no reference to what time it took to reach that, or that it was the same time for all three.

The way to benchmark RELION is to run it to the end, i.e. iteration 35 in the above case, and measure the wall time it took.

I'm building a intel version now to see if that makes any difference.

PS
See my comment in PR #8864 about 'usempi' toolchain option.

Reply:
Another thing, could you run the same benchmark without the --cpu flag?
Would be interesting to see what timings you get.

Reply:
And take note on what hardware you run the test on...

Reply:
@akesandgren : The 3 values are the iterations achieved with each version after the walltime. 
I will ask our CryoEM user to test it without the --cpu flag or actually provide a minimal example so we can benchmark it ourselves (walltime for all iterations).

Reply:
It should be fairly easy to reduce runtime, if it is the one i think it is.
In shiny_2sets.star (or whatever file is specified with --i just reduce the number of particle files listed. I cut it down to some 950, and have a runtime of ~20min with(out) --cpu using the intel build.

Reply:
@timeu FYI, GCC vs Intel doesn't make as much difference as openblas/FFTW3 vs MKL(fft)

Ok, one has to make a little patching in their CMake files to make it build with MKL and GCC but that's simple.

Going from foss to gomkl cuts time down to one third for the benchmark i'm running.

Reply:
Summary of reduced benchmark, running on a 28core E5-2690 v4 @ 2.60GHz, no MPI, --j 28
EB based foss default settings (no --cpu flag):33min
EB based foss default settings (--cpu): 1h04m
Handbuild foss, -O3 -ftree-vectorize threaded FFTW3 -DMKLFFT (--cpu): 1h06m
Handbuilt with gomkl, -O3 -ftree-vectorize -DFAST_CENTERFFT -DMKLFFT using mklfft-threaded (no --cpu): 24min
Handbuilt with gomkl, -O3 -ftree-vectorize -DFAST_CENTERFFT -DMKLFFT using mklfft-threaded (--cpu): 21min
EB based intel, opt, loose, (no --cpu): 19min
EB based intel, opt, loose, (--cpu): 13min

So using MKL is a BIG win for --cpu runs, cutting runtimes down to a third, using Intel instead of GCC cuts times down by another 50%.

So unless MKL is part of the mix, do not build ALT CPU code.